File: tests/python/aws-bedrock/test_prototype.py
"""Conformance test: prototype instrumentation for AWS Bedrock."""

import json
import os

from opentelemetry import trace
from opentelemetry._logs import get_logger_provider

from otel_setup import flush_and_shutdown, setup_otel

from common import run_memory_operations

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"]

_prototype_tracer = trace.get_tracer("gen_ai.prototype")


def create_bedrock_client():
    """Create a boto3 Bedrock Runtime client pointing at the mock server."""
    import boto3

    return boto3.client(
        "bedrock-runtime",
        endpoint_url=MOCK_BASE_URL,
        region_name="us-east-1",
        aws_access_key_id="mock",
        aws_secret_access_key="mock",
    )


def run_converse_prototype(client):
    """Scenario: Bedrock Converse API with prototype instrumentation."""
    print("  [converse] Bedrock Converse API (prototype)")
    request_model = "anthropic.claude-3-haiku-20240307-v1:0"
    with _prototype_tracer.start_as_current_span(
        "chat anthropic.claude-3-haiku-20240307-v1:0"
    ) as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "aws.bedrock")
        span.set_attribute("gen_ai.request.model", request_model)
        messages = [
            {
                "role": "user",
                "content": [{"text": "Say hello."}],
            }
        ]
        response = client.converse(
            modelId=request_model,
            messages=messages,
        )
        stop_reason = response.get("stopReason")
        if stop_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [stop_reason])
        usage = response.get("usage", {})
        if usage.get("inputTokens") is not None:
            span.set_attribute("gen_ai.usage.input_tokens", usage["inputTokens"])
        if usage.get("outputTokens") is not None:
            span.set_attribute("gen_ai.usage.output_tokens", usage["outputTokens"])
        text = response["output"]["message"]["content"][0]["text"]

        # Emit inference operation details event
        event_attrs = {
            "gen_ai.operation.name": "chat",
            "gen_ai.request.model": request_model,
            "gen_ai.input.messages": json.dumps([
                {"role": m["role"], "parts": [{"type": "text", "content": m["content"][0]["text"]}]}
                for m in messages
            ]),
            "gen_ai.output.messages": json.dumps([
                {
                    "role": "assistant",
                    "parts": [{"type": "text", "content": text}],
                    "finish_reason": stop_reason,
                }
            ]),
        }
        if stop_reason:
            event_attrs["gen_ai.response.finish_reasons"] = [stop_reason]
        if usage.get("inputTokens") is not None:
            event_attrs["gen_ai.usage.input_tokens"] = usage["inputTokens"]
        if usage.get("outputTokens") is not None:
            event_attrs["gen_ai.usage.output_tokens"] = usage["outputTokens"]
        get_logger_provider().get_logger("gen_ai.prototype").emit(
            event_name="gen_ai.client.inference.operation.details",
            body="Inference operation details",
            attributes=event_attrs,
        )

        print(f"    -> {text[:60]}")


def run_converse_tool_call_prototype(client):
    """Scenario: Bedrock Converse API with tool calling prototype instrumentation."""
    print("  [chat_tool_call] Bedrock Converse API with tool calling (prototype)")
    request_model = "anthropic.claude-3-haiku-20240307-v1:0"
    tool_spec = {
        "toolSpec": {
            "name": "get_weather",
            "description": "Get the current weather",
            "inputSchema": {
                "json": {
                    "type": "object",
                    "properties": {
                        "location": {"type": "string", "description": "City name"},
                    },
                    "required": ["location"],
                }
            },
        }
    }
    tool_config = {"tools": [tool_spec]}
    with _prototype_tracer.start_as_current_span(
        "chat anthropic.claude-3-haiku-20240307-v1:0"
    ) as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "aws.bedrock")
        span.set_attribute("gen_ai.request.model", request_model)
        span.set_attribute("gen_ai.tool.definitions", json.dumps(tool_config["tools"]))
        messages = [
            {
                "role": "user",
                "content": [{"text": "What's the weather in Seattle?"}],
            }
        ]
        response = client.converse(
            modelId=request_model,
            messages=messages,
            toolConfig=tool_config,
        )
        stop_reason = response.get("stopReason")
        if stop_reason:
            span.set_attribute("gen_ai.response.finish_reasons", [stop_reason])
        usage = response.get("usage", {})
        if usage.get("inputTokens") is not None:
            span.set_attribute("gen_ai.usage.input_tokens", usage["inputTokens"])
        if usage.get("outputTokens") is not None:
            span.set_attribute("gen_ai.usage.output_tokens", usage["outputTokens"])
        content = response["output"]["message"]["content"]
        if content and "toolUse" in content[0]:
            print(f"    -> tool_call: {content[0]['toolUse']['name']}")
        else:
            print(f"    -> {content[0]['text'][:60]}")


def run_embeddings_prototype(client):
    """Scenario: Bedrock Titan Embeddings with prototype instrumentation."""
    import json as _json

    print("  [embeddings] Bedrock Titan Embeddings (prototype)")
    request_model = "amazon.titan-embed-text-v2:0"
    with _prototype_tracer.start_as_current_span(
        "embeddings amazon.titan-embed-text-v2:0"
    ) as span:
        span.set_attribute("gen_ai.operation.name", "embeddings")
        span.set_attribute("gen_ai.provider.name", "aws.bedrock")
        span.set_attribute("gen_ai.request.model", request_model)
        response = client.invoke_model(
            modelId=request_model,
            contentType="application/json",
            accept="application/json",
            body=_json.dumps({"inputText": "Hello, world!"}),
        )
        result = _json.loads(response["body"].read())
        if result.get("inputTextTokenCount") is not None:
            span.set_attribute("gen_ai.usage.input_tokens", result["inputTextTokenCount"])
        print(f"    -> embedding dim: {len(result['embedding'])}")


def main():
    print("=== Prototype: AWS Bedrock Conformance Test ===")

    tp, lp, mp = setup_otel()

    client = create_bedrock_client()

    run_converse_prototype(client)
    run_converse_tool_call_prototype(client)
    run_embeddings_prototype(client)
    run_memory_operations(client)

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()

File: tests/js/aws-bedrock/test_prototype.ts
/**
 * Conformance test: Prototype instrumentation for AWS Bedrock.
 */
import { trace } from "@opentelemetry/api";
import { logs, SeverityNumber } from "@opentelemetry/api-logs";
import { flushAndShutdownOtel, setupOtel } from "../otel";
import { runMemoryOperations } from "./common";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_BASE_URL = nodeProcess.env.MOCK_LLM_URL!;
const tracer = trace.getTracer("gen_ai.prototype");

async function main() {
  console.log("=== Prototype: AWS Bedrock JS Conformance Test ===");
  const otel = setupOtel();

  const {
    BedrockRuntimeClient,
    ConverseCommand,
    ConverseStreamCommand,
    InvokeModelCommand,
  } = await import("@aws-sdk/client-bedrock-runtime");
  const { NodeHttpHandler } = await import("@smithy/node-http-handler");

  const client = new BedrockRuntimeClient({
    endpoint: MOCK_BASE_URL,
    region: "us-east-1",
    credentials: { accessKeyId: "mock", secretAccessKey: "mock" },
    requestHandler: new NodeHttpHandler(),
  });

  // Scenario: chat
  console.log("  [chat] basic chat completion (prototype)");
  await tracer.startActiveSpan("chat anthropic.claude-3-haiku-20240307-v1:0", async (span) => {
    const requestModel = "anthropic.claude-3-haiku-20240307-v1:0";
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "aws.bedrock");
    span.setAttribute("gen_ai.request.model", requestModel);
    span.setAttribute("gen_ai.response.model", requestModel);
    const messages = [
      { role: "user" as const, content: [{ text: "Say hello." }] },
    ];
    const resp = await client.send(
      new ConverseCommand({
        modelId: requestModel,
        messages,
      })
    );
    if (resp.stopReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [resp.stopReason]);
    }
    if (resp.usage) {
      if (resp.usage.inputTokens) span.setAttribute("gen_ai.usage.input_tokens", resp.usage.inputTokens);
      if (resp.usage.outputTokens) span.setAttribute("gen_ai.usage.output_tokens", resp.usage.outputTokens);
    }

    // Emit inference operation details event
    const outputText = resp.output?.message?.content?.[0]?.text ?? "";
    logs.getLogger("gen_ai.prototype").emit({
      severityNumber: SeverityNumber.INFO,
      eventName: "gen_ai.client.inference.operation.details",
      body: "Inference operation details",
      attributes: {
        "gen_ai.operation.name": "chat",
        "gen_ai.request.model": requestModel,
        "gen_ai.response.model": requestModel,
        "gen_ai.response.finish_reasons": resp.stopReason ? [resp.stopReason] : undefined,
        "gen_ai.usage.input_tokens": resp.usage?.inputTokens,
        "gen_ai.usage.output_tokens": resp.usage?.outputTokens,
        "gen_ai.input.messages": JSON.stringify(
          messages.map(m => ({ role: m.role, parts: [{ type: "text", content: m.content[0].text }] }))
        ),
        "gen_ai.output.messages": JSON.stringify([{
          role: "assistant",
          parts: [{ type: "text", content: outputText }],
          finish_reason: resp.stopReason,
        }]),
      },
    });

    console.log(`    -> ${outputText.slice(0, 60)}`);
    span.end();
  });

  // Scenario: chat with tool calling
  console.log("  [chat_tool_call] chat with tool calling (prototype)");
  await tracer.startActiveSpan("chat anthropic.claude-3-haiku-20240307-v1:0", async (span) => {
    const requestModel = "anthropic.claude-3-haiku-20240307-v1:0";
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "aws.bedrock");
    span.setAttribute("gen_ai.request.model", requestModel);
    span.setAttribute("gen_ai.response.model", requestModel);
    const toolSpec = {
      toolSpec: {
        name: "get_weather",
        description: "Get the current weather",
        inputSchema: {
          json: {
            type: "object",
            properties: {
              location: { type: "string", description: "City name" },
            },
            required: ["location"],
          },
        },
      },
    };
    const toolConfig = { tools: [toolSpec] };
    span.setAttribute("gen_ai.tool.definitions", JSON.stringify(toolConfig.tools));
    const messages = [
      { role: "user" as const, content: [{ text: "What's the weather in Seattle?" }] },
    ];
    const resp = await client.send(
      new ConverseCommand({
        modelId: requestModel,
        messages,
        toolConfig,
      })
    );
    if (resp.stopReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [resp.stopReason]);
    }
    if (resp.usage) {
      if (resp.usage.inputTokens) span.setAttribute("gen_ai.usage.input_tokens", resp.usage.inputTokens);
      if (resp.usage.outputTokens) span.setAttribute("gen_ai.usage.output_tokens", resp.usage.outputTokens);
    }
    const content = resp.output?.message?.content?.[0];
    if (content && "toolUse" in content && content.toolUse) {
      console.log(`    -> tool_call: ${content.toolUse.name}`);
    } else {
      console.log(`    -> ${content?.text?.slice(0, 60) ?? ""}`);
    }
    span.end();
  });

  // Scenario: streaming chat
  console.log("  [chat_streaming] streaming chat completion (prototype)");
  await tracer.startActiveSpan("chat anthropic.claude-3-haiku-20240307-v1:0", async (span) => {
    const requestModel = "anthropic.claude-3-haiku-20240307-v1:0";
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "aws.bedrock");
    span.setAttribute("gen_ai.request.model", requestModel);
    span.setAttribute("gen_ai.response.model", requestModel);
    const resp = await client.send(
      new ConverseStreamCommand({
        modelId: requestModel,
        messages: [
          { role: "user", content: [{ text: "Tell me a joke." }] },
        ],
      })
    );
    let text = "";
    let stopReason = "";
    let inputTokens = 0;
    let outputTokens = 0;
    if (resp.stream) {
      for await (const event of resp.stream) {
        if (event.contentBlockDelta?.delta?.text) {
          text += event.contentBlockDelta.delta.text;
        }
        if (event.messageStop?.stopReason) {
          stopReason = event.messageStop.stopReason;
        }
        if (event.metadata?.usage) {
          inputTokens = event.metadata.usage.inputTokens ?? 0;
          outputTokens = event.metadata.usage.outputTokens ?? 0;
        }
      }
    }
    if (stopReason) span.setAttribute("gen_ai.response.finish_reasons", [stopReason]);
    if (inputTokens) span.setAttribute("gen_ai.usage.input_tokens", inputTokens);
    if (outputTokens) span.setAttribute("gen_ai.usage.output_tokens", outputTokens);
    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  // Scenario: embeddings
  console.log("  [embeddings] Bedrock Titan Embeddings (prototype)");
  await tracer.startActiveSpan("embeddings amazon.titan-embed-text-v2:0", async (span) => {
    const requestModel = "amazon.titan-embed-text-v2:0";
    span.setAttribute("gen_ai.operation.name", "embeddings");
    span.setAttribute("gen_ai.provider.name", "aws.bedrock");
    span.setAttribute("gen_ai.request.model", requestModel);
    span.setAttribute("gen_ai.response.model", requestModel);
    const resp = await client.send(
      new InvokeModelCommand({
        modelId: requestModel,
        contentType: "application/json",
        accept: "application/json",
        body: JSON.stringify({ inputText: "Hello, world!" }),
      })
    );
    const result = JSON.parse(new TextDecoder().decode(resp.body));
    if (result.inputTextTokenCount) {
      span.setAttribute("gen_ai.usage.input_tokens", result.inputTextTokenCount);
    }
    console.log(`    -> embedding dim: ${result.embedding.length}`);
    span.end();
  });

  // Scenario: memory operations
  await runMemoryOperations();

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });
